916: seamless project import into eclipse

lostiniceland opened the pull request:
fixes bndtools/bndtools#916

Imports a bnd-workspace to Eclipse

Usage:

1. File -> Import...

1. BndTools -> Existing Bnd-Workspace

1. Choose a valid Bnd-Workspace root-folder

This importer will map informations from Bnd to Eclipse. To do so, **it will wipe all existing Eclipse project-specific settings in this workspace an recreate them**.

For each Bnd-project the import will make sure to set the correct source-, test-, and output-folders, where project level takes precedence over workspace-level settings.

Open
- Eclipse smart import (not sure if this is available in the current Eclipse 4.4.2 base)


This work is stalled since quite some time because I tried to display some more information in a read-only-panel (like mapped folders) but I wasnt very happy with this and couldnt find any better. So I just deleted this part.

Signed-off-by: Marc Schlegel <maschlegel@gmail.com>

gamerson commented:
Hey @lostiniceland 

Thanks for kicking off this effort.  This is very good feature that we would to see land in bndtools.  Could you help us with addressing some additional needs?  I've included the list you already mentioned plus some additional items.

- [ ] set execution environment based on bnd workspace model version
- [ ] Use Progress Monitor to show tasks (importing project foo, importing project bar, etc)
- [ ] add some type of bnd workspace model listener that will listen for key attributes of the workspace model (like osgi.ee level) and when it detects a change it will go through and propagate the necessary changes to the project models (update Java execution environment levels, etc).  Basically instead of make this a 'one off' type of operation which only happens on import, make it reactive as well.  
- [ ] provide some way for user to kick off this 'read workspace model' and propagate changes throughout.  Maybe an entry in the Bndtools > top level menu?
- [ ] regarding Eclipse smart import, we may be able to add a new bundle to bndtools that would have optional requirements, so if the user has Eclipse 4.6 or greater, this new bundle would become active and start contributing the extension point.  I can help with creating this if you are unsure of how to do it.

I have not looked at your code any yet, just the high level ideas.  I can wait for iteration #2 before I dive into code review.

lostiniceland commented:
Thanks for your ideas Gregory, I think it would be really helpfull if the workspace can handle certain changes in a reactive way and update itself. 

I've just updated the handling of the compiler-compliance according to the bnd-model. Going to work on the ProgressMonitor on the next day(s). I've added some TODOs where I am not exactly sure if what I did is right.

Though due to my lack of time I would like to propose to keep this PR to the level of importing a bnd-workspace statically and improve things later with the reactive features.

gamerson commented:
Agreed, just handling the import would be good enough, we could add more
improvements later.

On Mon, Feb 20, 2017 at 4:55 PM, Marc Schlegel <notifications@github.com>
wrote:

> Thanks for your ideas Gregory, I think it would be really helpfull if the
> workspace can handle certain changes in a reactive way and update itself.
>
> I've just updated the handling of the compiler-compliance according to the
> bnd-model. Going to work on the ProgressMonitor on the next day(s). I've
> added some TODOs where I am not exactly sure if what I did is right.
>
> Though due to my lack of time I would like to propose to keep this PR to
> the level of importing a bnd-workspace statically and improve things later
> with the reactive features.
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/bndtools/bndtools/pull/1653#issuecomment-281200847>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAkVFeV-KPRNwTFHt5FBLJPPgoqv7_NEks5rehnVgaJpZM4MCZUd>
> .
>



-- 
Greg Amerson
Liferay Developer Tools
Liferay, Inc. www.liferay.com

lostiniceland commented:
@gamerson could you please review and provide some feedback on the TODOs?

I've enhanced the error-handling, added some progress to the monitor, and figured out how to set the ExecutionEnvironment in the SystemLibray.
Regarding the EE, I am not sure if I should open a bug for Bnd since it looks to me we have no option to specifiy any ExecutionEnvironement during build (only run-framework). At the moment I am falling back to some fuzzy String-logic. Maybe something like "javac.ee: JavaSE-1.8"?

gamerson commented:
Hey Marc,

Yes I will give it a look later tonight!

On Tue, Feb 21, 2017 at 5:06 PM, Marc Schlegel <notifications@github.com>
wrote:

> @gamerson <https://github.com/gamerson> could you please review and
> provide some feedback on the TODOs?
>
> I've enhanced the error-handling, added some progress to the monitor, and
> figured out how to set the ExecutionEnvironment in the SystemLibray.
> Regarding the EE, I am not sure if I should open a bug for Bnd since it
> looks to me we have no option to specifiy any ExecutionEnvironement during
> build (only run-framework). At the moment I am falling back to some fuzzy
> String-logic. Maybe something like "javac.ee: JavaSE-1.8"?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/bndtools/bndtools/pull/1653#issuecomment-281512092>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAkVFYBh3BwopqHkm7ueTRLJT0JvyoBvks5re237gaJpZM4MCZUd>
> .
>



-- 
Greg Amerson
Liferay Developer Tools
Liferay, Inc. www.liferay.com

gamerson commented:
hey @lostiniceland sorry for the delay!  Here are some more review comments.

1. The spelling for Bndtools is just capitalize the first letter `Bndtools`
2. The thing that we are importing is `Bnd Workspace` so lets use that name 
3. I'm attached a new mockup of what we should be shooting for (modeled after existing standard eclipse import dialogs)

![importbndworkspace](https://cloud.githubusercontent.com/assets/595221/23277394/51e8aae2-f9d3-11e6-9bbb-ed02a7727be4.png)

So just a few things about this page:

1. first text field for root directory should not have an initial value (blank)
2. Should validate the user's selected directory
3. When user browses to a valid workspace project we should scan for projects that will be imported and list them in the `Projects` table with a checkbox list viewer where all projects by default are pre-selected
4. Then of course the deselect and select all should do what you would export, refresh would just refresh the list
5. take notice of the new label/strings and images (standard bndtools wizban that you can find used elsewhere)

bjhargrave commented:
> Then of course the deselect and select all should do what you would export, refresh would just refresh the list

I am not sure that is a good choice. Bnd does not care what project Eclipse knows about. Bnd care about the file systems and all the project are there on the file system. Failing to import some projects into Eclipse can cause errors if an imported project depends upon a non-imported project. So if we are going to allow partial import, then the wizard needs to check the project dependency relationships and make sure that for each imported project, all its dependency projects are also imported. See https://github.com/bndtools/bnd/blob/master/settings.gradle#L85-L92 for example gradle code for doing this.

gamerson commented:
OK, sure we could make it simpler and remove those buttons and not make that a checkbox table and just a plain list.

lostiniceland commented:
Thanks for the review guys.

Regarding the project selection: I remember Peter Kriens also mentioned once that the workspace should form a set of cohesive modules. Hence I decided against a partial import, reduced complexity being the other reason.

Another thing I came accross once Bndtools 3.3 was released with the new workspace-init (work started already with 3.2), was that when initializing a workspace via Bnd it immediately started pulling remote repos. Could this be? If true, would it be feasible to enhance Bndlib in a way that we can pull information from a workspace without initializing it? 

And one last thing: I would like to close the Wizard once the background-job is started, but I could not find the right API. The import runs in the background, but the wizard is frozen in the foreground. How do you think about this (is it possible at all with the Eclipse API)?

gamerson commented:
So I guess you can't get the projects (to make list) from Workspace without initing workspace?  Should we create a routine that does the equivalent of searching for all projects and showing them?  Or should we just remove the whole list of projects all together since it may be of dubious value at this point.

Regarding the "how to run a background job" and let the wizard close and have the import going on in the background, use this one.

```
new ProgressMonitorDialog( getShell() ).run( true, true, new IRunnableWithProgress()  {
  public void run( IProgressMonitor monitor ) throws InvocationTargetException, InterruptedException {
    monitor.beginTask( "Importing Bnd Workspace projects... ", 10 );
    monitor.done();
  }});
```

gamerson commented:
Also, I imported a small example set of projects that already had classpath entries and I wanted to see how the files were modified.  Here is the diff

https://gist.github.com/gamerson/434a3129dfbe2038544d97c7aaa176bc

I see a few things that I think we may want to change.

1. 
```
-	<classpathentry kind="src" output="bin" path="src"/>
+	<classpathentry kind="src" path="src"/>
```
 We need to keep the `output="bin"` attribute.  We thought about removing this before but decided it needed to stay this way so when we import we need to explicitly set the output path

2. Setting encoding to UTF-8 globally across the project?
```
-encoding/bnd.bnd=UTF-8
+encoding/<project>=UTF-8
```
I'm ok with that, but thought we might want to discuss.

3. whatever the java level is we should probably be setting these preferences.  Right now they are being removed, we should likely explicitly set them for each project.
```
-eclipse.preferences.version=1
-org.eclipse.jdt.core.compiler.codegen.inlineJsrBytecode=enabled
-org.eclipse.jdt.core.compiler.codegen.targetPlatform=1.8
-org.eclipse.jdt.core.compiler.codegen.unusedLocal=preserve
-org.eclipse.jdt.core.compiler.compliance=1.8
-org.eclipse.jdt.core.compiler.debug.lineNumber=generate
-org.eclipse.jdt.core.compiler.debug.localVariable=generate
-org.eclipse.jdt.core.compiler.debug.sourceFile=generate
-org.eclipse.jdt.core.compiler.problem.assertIdentifier=error
-org.eclipse.jdt.core.compiler.problem.enumIdentifier=error
-org.eclipse.jdt.core.compiler.source=1.8
```
4. A java nature is being added to the 'cnf' project.  I think we don't want to do this, just a normal 'general' project should be enough, no need for it to be a java project.
```
--- a/bndtools/cnf/.classpath
+++ b/bndtools/cnf/.classpath
@@ -1,6 +1,5 @@
 <?xml version="1.0" encoding="UTF-8"?>
 <classpath>
-	<classpathentry kind="src" path="src"/>
-	<classpathentry kind="con" path="org.eclipse.jdt.launching.JRE_CONTAINER"/>
+	<classpathentry kind="con" path="org.eclipse.jdt.launching.JRE_CONTAINER/org.eclipse.jdt.internal.debug.ui.launcher.StandardVMType/JavaSE-1.8"/>
 	<classpathentry kind="output" path="bin"/>
 </classpath>
diff --git a/bndtools/cnf/.project b/bndtools/cnf/.project
index 1397bbe..0b71642 100644
--- a/bndtools/cnf/.project
+++ b/bndtools/cnf/.project
@@ -5,7 +5,13 @@
 	<projects>
 	</projects>
 	<buildSpec>
+		<buildCommand>
+			<name>org.eclipse.jdt.core.javabuilder</name>
+			<arguments>
+			</arguments>
+		</buildCommand>
 	</buildSpec>
 	<natures>
+		<nature>org.eclipse.jdt.core.javanature</nature>
 	</natures>
```

lostiniceland commented:
I've updated the UI according to @gamerson comments. Let me know if I forgot something.

Regarding the last comment.

1. The output="bin" (or whatever is specified in Bnd) is only applied if the folder is different from the Eclipse workspace settings (see [importSourceAndOutputFolders](https://github.com/bndtools/bndtools/pull/1653/files#diff-93b614df32b572342b5877638cd44bb2R257))

1. Same for the UTF-8. The import doesn't set any encoding by itself. The changed encoding is probably coming from the configured Eclipse workspace setting. 

1. Do we really need those JDT settings (afaik there is no API in Eclipse to set these kind of properties)? The current setup works as far as I can tell, and I am able to set different JDK targets (even within one Bnd Workspace). 

1. Java-Nature was removed in the last commit. You might want to update the current Bndtools templates which are used to create a new workspace. Because there, the cnf is a java-project.

As you see, most changes in the [diff ](https://gist.github.com/gamerson/434a3129dfbe2038544d97c7aaa176bc) are probably related to the different way the workspace is created compared to the "New Bnd Workspace Wizard". Whereas the "New Bnd Workspace Wizard" uses file-templates, this PR makes use of public API which has the benefit that global settings are also incorporated.

gamerson commented on bndtools.core/src/bndtools/wizards/project/ImportBndWorkspaceWizard.java:
```diff
@@ -0,0 +1,451 @@
+    @Override
+    public boolean performFinish() {
+        boolean result = false;
+        if (mainPage.isPageComplete()) {
```
Do we need to check this?  I believe that `performFinish` should not get called unless all pages are `complete`.

gamerson commented on bndtools.core/src/bndtools/wizards/project/ImportBndWorkspaceWizard.java:
```diff
@@ -0,0 +1,451 @@
+
+        // create generic project
+        final IProjectDescription cnfProjectDescription = eclipseWorkspace.newProjectDescription(Workspace.CNFDIR);
+        final IProject project = eclipseWorkspaceRoot.getProject(Workspace.CNFDIR);
```
Does this code handle configurable cnf directory paths?

lostiniceland replied:
No, I think not. Will look at the Bnd-Api...

gamerson commented on bndtools.core/src/bndtools/wizards/project/ImportBndWorkspaceWizard.java:
```diff
@@ -0,0 +1,451 @@
+        final IProjectDescription cnfProjectDescription = eclipseWorkspace.newProjectDescription(Workspace.CNFDIR);
+        final IProject project = eclipseWorkspaceRoot.getProject(Workspace.CNFDIR);
+
+        deleteOldProjectFiles(Paths.get(bndWorkspace.getBase().toURI()).resolve(Workspace.CNFDIR));
```
NOw that I think of it, should we be deleteing existing project files?  What if they have other natures that are important to their project.  I think that we should leave existing project files and just augment them.

lostiniceland replied:
There were some problems with the project-settings when old one are not deleted. For example the classpath was messed up.
Just give it a try with this method commented out and have a look at the generated projects (build path). 

If we can solve this, I am with you that the settings should be kept. I could add a checkbox "clear old settings" because sometimes it nice to start fresh.

gamerson commented on bndtools.core/src/bndtools/wizards/project/ImportBndWorkspaceWizard.java:
```diff
@@ -0,0 +1,451 @@
+            // Now the created source folder should be added to the class entries of the project, otherwise compilation will fail
+            IPackageFragmentRoot root = javaProject.getPackageFragmentRoot(source);
+            List<IClasspathEntry> entries = new ArrayList<>(Arrays.asList(javaProject.getRawClasspath()));
+            entries.add(JavaCore.newSourceEntry(root.getPath(), null, defaultOutputLocation ? null : outputRoot.getPath()));
```
This sourceEntry we need to add an explicitly output path to `bin` just like the `New Bnd Project` creates here:  https://gist.github.com/gamerson/710a7be00790bb9b74be04c9867f64d2#file-diff-txt-L8

lostiniceland replied:
But what if the project uses a maven-layout? Or do you mean that if the Bnd Workspace doesnt specify anything it has to be "bin" regardless what the Eclipse preference say?

gamerson commented on bndtools.core/src/bndtools/wizards/project/ImportBndWorkspaceWizard.java:
```diff
@@ -0,0 +1,451 @@
+     */
+    private void addSystemLibraryContainer(IJavaProject javaProject, String javacTarget, IProgressMonitor monitor) throws JavaModelException {
+        // TODO resolution of ExecutionEnvironment sketchy. Which Bnd property should be used? It seems theres a BncConstant missing for setting an EE globally
+        IExecutionEnvironment environment = JavaRuntime.getExecutionEnvironmentsManager().getEnvironment("JavaSE-" + javacTarget);
```
Lets just check the BndConstants.JAVAC_SOURCE or BndConstants.JAVAC_TARGET property if they exist.

gamerson commented on bndtools.core/src/bndtools/wizards/project/ImportBndWorkspaceWizard.java:
```diff
@@ -0,0 +1,451 @@
+    }
+
+    @Override
+    public boolean canFinish() {
```
Can we remove this method?

gamerson commented on bndtools.core/src/bndtools/wizards/project/ImportBndWorkspaceWizardPageOne.java:
```diff
@@ -0,0 +1,249 @@
+                    bndWorkspace = Workspace.getWorkspace(chosenDirectory);
+                    setErrorMessage(null);
+                    txtFolderErrorDecorator.hide();
+                    tableViewer.setInput(bndWorkspace.getAllProjects());
```
Lets add the `cnf` folder to the tableViewer so we may need to add more than just bndWorkspace.getAllProjects() since that doesn't include the `cnf` project.

gamerson replied:
Can we try to add the CNF here?

gamerson commented on bndtools.core/src/bndtools/wizards/project/ImportBndWorkspaceWizard.java:
```diff
@@ -0,0 +1,451 @@
+        project.create(cnfProjectDescription, monitor);
+        project.open(monitor);
+
+        project.setDefaultCharset("UTF-8", monitor);
```
If this is a new project lets set the charset to UTF-8.  But if this is an existing project (already have a .project) we should probably leave it so we don't wipe out the user's charset settings.

lostiniceland replied:
Woops, missed that one on my last reply regarding the encoding :-)

gamerson commented:
Agreed with the 'clear old project metadata' option.  We could give users
the option to wipe out and start fresh but we should make another attempt
at preserving the existing meta.  I know that the m2e ran into this, and
users had the expectation that existing .settings/ and .project/.classpath
would try to be honored where it could be.

On Thu, Mar 16, 2017 at 12:13 PM, Marc Schlegel <notifications@github.com>
wrote:

> *@lostiniceland* commented on this pull request.
> ------------------------------
>
> In bndtools.core/src/bndtools/wizards/project/
> ImportBndWorkspaceWizard.java
> <https://github.com/bndtools/bndtools/pull/1653#discussion_r106477320>:
>
> > +    }
> +
> +    private void importConfigurationProject(File rootImportPath, IProgressMonitor monitor) throws Exception {
> +        monitor.subTask("Import configuration project 'cnf'.");
> +        final IWorkspace eclipseWorkspace = ResourcesPlugin.getWorkspace();
> +        final IWorkspaceRoot eclipseWorkspaceRoot = eclipseWorkspace.getRoot();
> +        final Workspace bndWorkspace = Workspace.getWorkspace(rootImportPath);
> +
> +        // Prepare Eclipse Workspace
> +        updateEclipseWorkspaceSettings(bndWorkspace);
> +
> +        // create generic project
> +        final IProjectDescription cnfProjectDescription = eclipseWorkspace.newProjectDescription(Workspace.CNFDIR);
> +        final IProject project = eclipseWorkspaceRoot.getProject(Workspace.CNFDIR);
> +
> +        deleteOldProjectFiles(Paths.get(bndWorkspace.getBase().toURI()).resolve(Workspace.CNFDIR));
>
> There were some problems with the project-settings when old one are not
> deleted. For example the classpath was messed up.
> Just give it a try with this method commented out and have a look at the
> generated projects (build path).
>
> If we can solve this, I am with you that the settings should be kept. I
> could add a checkbox "clear old settings" because sometimes it nice to
> start fresh.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/bndtools/bndtools/pull/1653#discussion_r106477320>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAkVFaSywgoBYH7cOFiLXkunlXFQDLMZks5rmW3UgaJpZM4MCZUd>
> .
>



-- 
Greg Amerson
Liferay Developer Tools
Liferay, Inc. www.liferay.com

gamerson commented:
If there are existing classpath settings we could honor them.  Its just for
new projects we should set it.

On Thu, Mar 16, 2017 at 4:40 PM, Marc Schlegel <notifications@github.com>
wrote:

> *@lostiniceland* commented on this pull request.
> ------------------------------
>
> In bndtools.core/src/bndtools/wizards/project/
> ImportBndWorkspaceWizard.java
> <https://github.com/bndtools/bndtools/pull/1653#discussion_r106538401>:
>
> > +        // remove defaults
> +        removeClasspathDefaults(javaProject);
> +
> +        // Output
> +        IFolder sourceOutput = workspaceProject.getFolder(URIUtil.toPath(bndProject.getSrcOutput().toURI()).makeRelativeTo(workspaceProject.getLocation()));
> +        IPackageFragmentRoot outputRoot = javaProject.getPackageFragmentRoot(sourceOutput);
> +
> +        final boolean defaultOutputLocation = outputRoot.getElementName().equals(javaProject.getOutputLocation().makeRelativeTo(javaProject.getPath()).toString());
> +
> +        // Source (multiple possible)
> +        for (File folder : bndProject.getSourcePath()) {
> +            IFolder source = workspaceProject.getFolder(URIUtil.toPath(folder.toURI()).makeRelativeTo(workspaceProject.getLocation()));
> +            // Now the created source folder should be added to the class entries of the project, otherwise compilation will fail
> +            IPackageFragmentRoot root = javaProject.getPackageFragmentRoot(source);
> +            List<IClasspathEntry> entries = new ArrayList<>(Arrays.asList(javaProject.getRawClasspath()));
> +            entries.add(JavaCore.newSourceEntry(root.getPath(), null, defaultOutputLocation ? null : outputRoot.getPath()));
>
> But what if the project uses a maven-layout? Or do you mean that if the
> Bnd Workspace doesnt specify anything it has to be "bin" regardless what
> the Eclipse preference say?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/bndtools/bndtools/pull/1653#discussion_r106538401>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAkVFQSXNI-BEUpdq_j-5URFz5-O1ImPks5rmaxBgaJpZM4MCZUd>
> .
>



-- 
Greg Amerson
Liferay Developer Tools
Liferay, Inc. www.liferay.com

gamerson commented:
hey @lostiniceland  

I know it may take quite a bit of effort to support keeping existing .settings/ .project and .classpath.  Lets move that support to another PR if we find we want it.  This use-case is for people importing a brand new bnd workspace that has never been in eclipse before and this solves that.  Once users have imported everything and if they want to customize settings and natures, etc, then next time they import to a new eclipse they can just use normal eclipse importer.

So can we do one more thing?  Since we are going to delete all existing .project/.settings can we at least prompt the user's first?  So they have to give the OK before we wipe existing files?  Once we do that, I think we could merge this.

gamerson commented on bndtools.core/src/bndtools/wizards/project/ImportBndWorkspaceWizard.java:
```diff
@@ -0,0 +1,438 @@
+
+        // create BndToolsProject
+        // TODO why create BndProject?
+        BndProject bndtoolsProject = new BndProject(javaProject.getProject());
```
Is this needed to make it work?

lostiniceland replied:
I dont know :-) In my tests I did not but I wasnt sure if this might trigger something else.
Will remove it

lostiniceland commented:
I am working on adding the cnf to the viewer right now...doing some additional decorations.

I've already been doing some investigations on how to preserve the settings, I hope this can be fixed within the next week.

One more question regarding the name of the cnf-project. As far as I remember this invariant was carved in stone (it has to be named cnf). Could you provide me with a snippet how I can handle this with bndlib?

gamerson commented:
Looks like there is a way according to this:
http://bnd.bndtools.org/chapters/150-build.html

But I think perhaps its best to just hard code it to 'cnf' for now and we
can improve it with the configurable 'cnf' version.

On Fri, Mar 17, 2017 at 4:55 PM, Marc Schlegel <notifications@github.com>
wrote:

> I am working on adding the cnf to the viewer right now...doing some
> additional decorations.
>
> I've already been doing some investigations on how to preserve the
> settings, I hope this can be fixed within the next week.
>
> One more question regarding the name of the cnf-project. As far as I
> remember this invariant was carved in stone (it has to be named cnf). Could
> you provide me with a snippet how I can handle this with bndlib?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/bndtools/bndtools/pull/1653#issuecomment-287481220>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAkVFTt-5WQhQ_niipxQaCoaY-Vb2HC7ks5rmwFtgaJpZM4MCZUd>
> .
>



-- 
Greg Amerson
Liferay Developer Tools
Liferay, Inc. www.liferay.com

lostiniceland commented:
@gamerson I think we are ready for another review-iteration.

I've readded the execution-environment for eclipse, but made it optional. It will infer "J2SE" and "JavaSE" environments from javac.target setting in Bnd, but only if there isnt already a existing JRE-container. Though not perfect yet due to missing environments I think it is handy for many projects which use Standard-Java.

Furthermore I did some tests with existing settings and the result seems ok. I had to keep [this method](https://github.com/lostiniceland/bndtools/blob/916-import-project/bndtools.core/src/bndtools/wizards/project/ImportBndWorkspaceWizard.java#L398-L405), but this did not have any side effects on the project-settings (I guess it does if you add a custom source-folder which is not known in Bnd, but that doesn make any sense in a Bnd Workspace).

Regarding the configurable cnf-location...it should already work but there seems to be a bug in Bnd. Maybe @bjhargrave can give some information about this: I tested the following with the provided **bndws** workspace

1. renamed cnf folder to xyz

1. created a file cnf which had the content xyz

1. import project, which calls Workspace.getWorkspace(path to bndws)

Exception: no workspace found under .../bndws/xyz
It seems Bnd is interpreting content of the cnf file as bnd-workspace-root rather than the location of a cnf-folder. According to section 8.1 in the [Bnd documentation](http://bnd.bndtools.org/chapters/150-build.html)) the contents of a cnf-file are used to find the cnf-folder.

gamerson commented on bndtools.core/src/bndtools/wizards/project/ImportBndWorkspaceWizardPageOne.java:
```diff
@@ -0,0 +1,378 @@
+        deleteButton = new Button(container, SWT.CHECK);
+        deleteButton.setText("Delete existing settings");
+
+        inferExecutionEnvironmentButton = new Button(container, SWT.CHECK);
```
how about we make this button checked by default?

lostiniceland replied:
Done

gamerson commented:
Thanks for the feature! @lostiniceland
